### PR TITLE
protocol: add payload compression

### LINKED ISSUE #872: Implement message flags and lz4 compression
We need some lz4 library for that. PayloadMaxSize limit should also be implemented there. One thing I'd like to change from C# implementation is make it a little less aggressive in compression, I think for everything that roughly fits into one packet compressing doesn't make much sense and also there are payloads that don't compress much (like `HeadersPayload`), so we need to account for both factors when deciding whether to compress or not. So, I propose these heuristics:
 * use 1024 for `CompressionMinSize`
 * don't use `CompressionThreshold`
 * don't even try to compress `HeadersPayload`, `MerkleBlockPayload`

### PR DESCRIPTION
**Pull request by AnnaShaleva**
closes #872

### CODE REVIEW

File: pkg/io/compressor.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	"io"
+
+	"github.com/pierrec/lz4"
+)
+
+// Compress compresses bytes using lz4.
+func Compress(source []byte) ([]byte, error) {
+	dest := NewBufBinWriter()
+	w := lz4.NewWriter(dest.w)
+	_, err := io.Copy(w, NewBinReaderFromBuf(source).r)
```
Comment (roman-khimov) [inline pkg/io/compressor.go:13]: Same thing here with `BinReader`.

File: pkg/network/message.go [hunk 2]
```diff
@@ -12,18 +13,38 @@ import (
 
 //go:generate stringer -type=CommandType
 
+const (
+	// PayloadMaxSize is maximum payload size in decompressed form.
+	// It is set to be 2000000 in hexadecimal representation.
+	PayloadMaxSize = 33554432
```
Comment (roman-khimov) [inline pkg/network/message.go:19]: So why not use hex representation for it?

File: pkg/network/message.go [hunk 3]
```diff
@@ -147,11 +159,11 @@ func (m *Message) decodePayload(br *io.BinReader) error {
 
 // Encode encodes a Message to any given BinWriter.
 func (m *Message) Encode(br *io.BinWriter) error {
+	m.tryCompressPayload()
```
Comment (roman-khimov) [inline pkg/network/message.go:162]: I think you should either have it here or in `NewMessage`, but not in both.

File: pkg/network/message.go [hunk 4]
```diff
@@ -170,3 +182,32 @@ func (m *Message) Bytes() ([]byte, error) {
...
+
+func (m *Message) tryCompressPayload() {
+	if m.Flags&Compressed == 0 && m.Payload != nil && m.compressedPayload == nil {
+		buf := io.NewBufBinWriter()
+		m.Payload.EncodeBinary(buf.BinWriter)
+		if buf.Err != nil {
+			panic(buf.Err)
+		}
+		compressedPayload := buf.Bytes()
+		size := uint32(len(compressedPayload))
```
Comment (roman-khimov) [inline pkg/network/message.go:194]: I don't think you need `uint32` here.

File: pkg/io/compressor.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package io
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/io/compressor.go]: Moved it to `message`

File: pkg/network/message.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/network/message.go]: Removed this method from `NewMessage`

File: pkg/network/message.go [hunk 6]
```diff
@@ -1,29 +1,52 @@
 package network
 
 import (
+	"bytes"
+	"errors"
 	"fmt"
+	iocore "io"
```
Comment (roman-khimov) [inline pkg/network/message.go:7]: You can move compress/decompress functions into their own `compress.go` here and avoid doing this.

File: pkg/network/message.go [hunk 7]
```diff
@@ -65,46 +88,37 @@ const (
...
+	m.compressedPayload = br.ReadVarBytes()
+	if len(m.compressedPayload) > PayloadMaxSize {
+		return errors.New("invalid payload size")
 	}
-	// return if their is no payload.
-	if m.Length == 0 {
-		return nil
-	}
-	return m.decodePayload(br)
+	return m.decodePayload()
```
Comment (roman-khimov) [inline pkg/network/message.go:106]: You need to check `br.Err` before doing this. This is not a `DecodeBinary`, this method returns an error and it should return a proper one for any kind of message given.

File: pkg/network/message_test.go [hunk 8]
```diff
@@ -1 +1,32 @@
...
+import (
+	"testing"
+	"time"
+
+	"github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/network/payload"
+	"github.com/stretchr/testify/require"
+)
+
+func TestEncodeDecode(t *testing.T) {
```
Comment (roman-khimov) [inline pkg/network/message_test.go:12]: I think we need more tests to cover this changeset.

Comment (AnnaShaleva) [reply pkg/network/message_test.go]: I added test for boundary cases to check whether compression works correctly. Do we need tests for all message types?

Comment (roman-khimov) [reply pkg/network/message_test.go]: I think it's enough to cover changes made by this PR which is mostly done as of now.
